machine-learning/lambda-v2/create_components.py

```diff
@@ -0,0 +1,369 @@
+            for file in os.listdir(version_path):
+                if file.startswith("."):
+                    continue
+                file_path = os.path.join(version_path, file)
+                build_file = os.path.join(build, file)
+                if os.path.isdir(file_path):
+                    shutil.make_archive(build_file, "zip", file_path)
+                    zip_file = "{}.zip".format(build_file)
+                    rel_path = os.path.relpath(zip_file, build_dir_path)
+                    upload_artifacts(zip_file, rel_path)
+                elif os.path.isfile(file_path):
+                    rel_path = os.path.relpath(build_file, build_dir_path)
+                    shutil.copy(file_path, build_file)
+                    upload_artifacts(build_file, rel_path)
+            if ".Model" in component:
+                download_model(build, inference_models[component])
+
+def create_lambda_artifacts(component_arg):
+    for component in os.listdir(lambda_artifacts_path):
+        if (
+            (component_arg != "" and component_arg != component)
+            or component.startswith(".")
+            or inferenceType not in component
+        ):
+            continue
+        component_path = os.path.join(lambda_artifacts_path, component)
+        for version in os.listdir(component_path):
+            if version.startswith("."):
+                continue
+            version_path = os.path.join(component_path, version)
+            component_latest_version = get_latest_version(component, version)
+            build = os.path.join(build_artifacts_path, component, component_latest_version)
+            os.makedirs(build, mode=0o777, exist_ok=True)
+            for file in os.listdir(version_path):
+                if file.startswith("."):
+                    continue
+                file_path = os.path.join(version_path, file)
+                build_file = os.path.join(build, file)
+                if os.path.isdir(file_path):
+                    shutil.make_archive(build_file, "zip", file_path)
+                    zip_file = "{}.zip".format(build_file)
+                    rel_path = os.path.relpath(zip_file, build_dir_path)
+                    create_lambda(zip_file, component)
+
+def create_lambda(zip_file, component):
+    function_name = component.split(".")[2]
+    if lambda_role == "":
+        print("Please pass in the lambda execution role required to create the inference lambda as '--lambdaRole' or '-l' argument")
+        exit(1)
+    try:
+        response = lambda_client.create_function(
+            FunctionName=function_name,
+            Runtime ='python3.8',
+            Role = "arn:aws:iam::{}:role/{}".format(get_account_number(), lambda_role),
+            Handler="inference.lambda_handler",
+            Timeout=123,
+            MemorySize=128,
+            Publish=True,
+            PackageType='Zip',
+            Code={
+            'ZipFile': open(zip_file, 'rb').read()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *open*. The resource is allocated by call *builtins.open*. Execution paths that do not contain closure statements were detected. To prevent this resource leak, close the object returned by *open()* in a try-finally block or declare it using a `with` statement.

**More info**
[View details about the `with` statement in the Python developer&#39;s guide](https://www.python.org/dev/peps/pep-0343/) (external link).